server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -289,9 +293,21 @@ public RecoveryResponse newInstance() {
      * @param recoveryTarget the target of the recovery
      * @return a snapshot of the store metadata
      */
-    private Store.MetadataSnapshot getStoreMetadataSnapshot(final RecoveryTarget recoveryTarget) {
+    static Store.MetadataSnapshot getStoreMetadataSnapshot(final Logger logger, final RecoveryTarget recoveryTarget) {
         try {
-            return recoveryTarget.indexShard().snapshotStoreMetadata();
+            final Store.MetadataSnapshot snapshot = recoveryTarget.indexShard().snapshotStoreMetadata();
+            // The primary shard may need the "exact" numDocs to verify if the commit has syncId.
+            final boolean softDeleteEnabled = recoveryTarget.indexShard().indexSettings().isSoftDeleteEnabled();
+            if (softDeleteEnabled && Strings.hasText(snapshot.getSyncId())) {
+                final List<IndexCommit> commits = DirectoryReader.listCommits(recoveryTarget.store().directory());
```

Comment:
This needs to use safe commit. This is the commit the engine will use (probably an existing bug) .